### PR TITLE
Issue #1859: Hot fix for metadata

### LINKED ISSUE #1859: The meta data report is not able to download on stable
### Current Behavior

The meta data report is not able to download on stable and results in an error

### Expected Behavior

The meta data report should download

### Steps to Reproduce

1.go to manage.get.gov/admin/registrar/domain
2. click the meta data report
3. see it takes for ever to load and throws an error

### Environment

stable

works fine on others

### Additional Context

Adding additional columns on the meta data report has resulted in a lot more computation time when it comes to our full data set. Look at the logic around columns and you will see some duplication that could be reduced. 

### Issue Links

Relates to #1840 and #1556

### PR DESCRIPTION
**Pull request by rachidatecs**
## Ticket

Resolves #1859

## Changes

- Refactor how we get the max domain managers count (on the domain that has the most)

#### Satisfied acceptance criteria and met development standards

- [x] Met the acceptance criteria, or will meet them in a subsequent PR
- [x] Created/modified automated tests
- [x] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [x] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [ ] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
- [ ] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

### CODE REVIEW

File: src/registrar/utility/csv_export.py
Comment (therealslimhsiehdy) [inline src/registrar/utility/csv_export.py:170]: **Question:** Just to confirm -- because we're not doing the domain_info call outside, and bc we already have it and it's within the loop that helps the time intensiveness, correct?

File: src/registrar/utility/csv_export.py
Comment (therealslimhsiehdy) [inline src/registrar/utility/csv_export.py:182]: **Question:** To confirm -- we're also getting the row count and extending it once versus writing the row each time here, correct?

File: src/registrar/utility/csv_export.py
Comment (rachidatecs) [reply src/registrar/utility/csv_export.py:173]: That's the hope

File: src/registrar/utility/csv_export.py
Comment (therealslimhsiehdy) [inline src/registrar/utility/csv_export.py:221]: Don't forget to remove this when you merge 😆

File: src/registrar/utility/csv_export.py
Comment (rachidatecs) [reply src/registrar/utility/csv_export.py:185]: Yeah. That's not a good thing. That's needed because we can't write body rows before we write the header. We can't write the header before we know max-dm. We can't know max_dm before we iterate over all DIs. We get the row data while iterating over DIs. So, we hold off on writing and store them all in a mega-list until the header is done.

File: src/registrar/utility/csv_export.py
Comment (rachidatecs) [reply src/registrar/utility/csv_export.py:221]: DUDE!

File: src/registrar/utility/csv_export.py
Comment (therealslimhsiehdy) [inline src/registrar/utility/csv_export.py:208]: I'M SORRY ITS ME AGAIN I think for testing purposes you may have commented it out, but I think we want to keep the security contact email for this data collection (I might be missing context so lmk if I'm incorrect) 😅

File: src/registrar/utility/csv_export.py
Comment (abroddrick) [inline src/registrar/utility/csv_export.py:50]: **non blocking suggestion**: this is for a section not part of your changes
This part below will just cause an unhelpful 500 error to show on screen. Fine for now as hopefully this should never happen but bad practice, we should at least be logging the error before throwing errors. 

Mind adding a log there? Not
```
 if domain_info.domain is None:
        raise ValueError("Domain is none")
```

File: src/registrar/utility/csv_export.py
Comment (abroddrick) [inline src/registrar/utility/csv_export.py:100]: **non blockingsuggestion**
Hey so won't we ONLY need to add to this list if this check in write_csv actually enters and updates the max dm count values " if dm_count > max_dm_count:
                max_dm_count = dm_count"

Instead of always doing this call I would think just passing an update_domain_count bool could would save on looping through the columns. line 102 has O(n) complexity. Not ideal, but also this shouldn't hurt a ton given that list shouldn't be too big.

File: src/registrar/utility/csv_export.py
Comment (abroddrick) [reply src/registrar/utility/csv_export.py:173]: I think I made a very similiar suggestion to zander's here, but on the actual loop in question. and I think moving it here would save on having a bool passed to the other function.

File: src/registrar/utility/csv_export.py
Comment (abroddrick) [inline src/registrar/utility/csv_export.py:184]: **comment** don't get why there is a continue here, python continues on excepted errors anyways

File: src/registrar/utility/csv_export.py
Comment (abroddrick) [reply src/registrar/utility/csv_export.py:185]: So there is a difference on writre row vs rows but it's dependent on the amount of data we have. Memory wise, yes storing all rows in memory is costly on memory but not on computation time. Whereas as writerow on large data sets is more expensive in time but less expensive in memory. Time is kind of our killer here and is really dependent on the large data set. A timer on the methods would require a large data set to get an accurate view, but the documentation online on row vs rows is a bit more enlightening. Writerows seems like the better call based on online searches

**Review by abroddrick — COMMENTED** [review]
Almost done reviewing no red flags so far just want to share in case I don't finish before our next meeting

File: src/registrar/utility/csv_export.py
Comment (abroddrick) [reply src/registrar/utility/csv_export.py:208]: **must change before merging** I agree looking at staging the security email is supposed to be here. Add it back in before merging

**Review by abroddrick — APPROVED** [review]
This seems to work, I made several comments, only one of which is critical and needs to address before merging. Approving with the understanding that one will at least be changed OR you will tell me I am wrong. Consider the other ones, but not going to block a hotfix for them.

**Pull request merged**
